File: ts/tests/config.test.ts
import { loadCases } from './helpers';
import { senderConfigFromEnv, botConfigFromEnv, receiverConfigFromEnv } from '../src/config';

interface ConfigCases {
  sender: {
    valid: { env: Record<string, string>; expected: { webex_token: string; target: string; slug: string; domains: string[]; debug: boolean } };
    missing_token: { env: Record<string, string>; error_contains: string };
    missing_target: { env: Record<string, string>; error_contains: string };
    debug_defaults_false: { env: Record<string, string>; expected_debug: boolean };
    domains_optional: { env: Record<string, string>; expected_domains: string[] };
  };
  bot: {
    valid: { env: Record<string, string>; expected: { webex_token: string; domains: string[] } };
    missing_domains: { env: Record<string, string>; error_contains: string };
  };
  receiver: {
    valid: { env: Record<string, string>; expected: { webex_token: string; slug: string; domains: string[] } };
  };
  debug_truthy_values: string[];
  debug_falsy_values: string[];
}

const CASES = loadCases<ConfigCases>('config_cases.json');

function setEnv(env: Record<string, string>): void {
  // Clear WGROK_ vars
  for (const key of Object.keys(process.env)) {
    if (key.startsWith('WGROK_')) {
      delete process.env[key];
    }
  }
  for (const [k, v] of Object.entries(env)) {
    process.env[k] = v;
  }
}

afterEach(() => {
  for (const key of Object.keys(process.env)) {
    if (key.startsWith('WGROK_')) {
      delete process.env[key];
    }
  }
});

describe('senderConfigFromEnv', () => {
  it('valid', () => {
    setEnv(CASES.sender.valid.env);
    const cfg = senderConfigFromEnv();
    expect(cfg.webexToken).toBe(CASES.sender.valid.expected.webex_token);
    expect(cfg.target).toBe(CASES.sender.valid.expected.target);
    expect(cfg.slug).toBe(CASES.sender.valid.expected.slug);
    expect(cfg.domains).toEqual(CASES.sender.valid.expected.domains);
    expect(cfg.debug).toBe(CASES.sender.valid.expected.debug);
  });

  it('missing token', () => {
    setEnv(CASES.sender.missing_token.env);
    expect(() => senderConfigFromEnv()).toThrow(new RegExp(CASES.sender.missing_token.error_contains, 'i'));
  });

  it('missing target', () => {
    setEnv(CASES.sender.missing_target.env);
    expect(() => senderConfigFromEnv()).toThrow(new RegExp(CASES.sender.missing_target.error_contains, 'i'));
  });

  it('debug defaults false', () => {
    setEnv(CASES.sender.debug_defaults_false.env);
    expect(senderConfigFromEnv().debug).toBe(CASES.sender.debug_defaults_false.expected_debug);
  });

  it('domains optional', () => {
    setEnv(CASES.sender.domains_optional.env);
    expect(senderConfigFromEnv().domains).toEqual(CASES.sender.domains_optional.expected_domains);
  });
});

describe('botConfigFromEnv', () => {
  it('valid', () => {
    setEnv(CASES.bot.valid.env);
    const cfg = botConfigFromEnv();
    expect(cfg.webexToken).toBe(CASES.bot.valid.expected.webex_token);
    expect(cfg.domains).toEqual(CASES.bot.valid.expected.domains);
  });

  it('missing domains', () => {
    setEnv(CASES.bot.missing_domains.env);
    expect(() => botConfigFromEnv()).toThrow(new RegExp(CASES.bot.missing_domains.error_contains, 'i'));
  });
});

describe('receiverConfigFromEnv', () => {
  it('valid', () => {
    setEnv(CASES.receiver.valid.env);
    const cfg = receiverConfigFromEnv();
    expect(cfg.webexToken).toBe(CASES.receiver.valid.expected.webex_token);
    expect(cfg.slug).toBe(CASES.receiver.valid.expected.slug);
    expect(cfg.domains).toEqual(CASES.receiver.valid.expected.domains);
  });
});

File: rust/tests/config_test.rs
use serde::Deserialize;
use std::collections::HashMap;
use std::fs;
use std::sync::Mutex;
use wgrok::config::{parse_debug, BotConfig, ReceiverConfig, SenderConfig};

static ENV_LOCK: Mutex<()> = Mutex::new(());

#[derive(Deserialize)]
struct ConfigCases {
    sender: SenderCases,
    bot: BotCases,
    receiver: ReceiverCases,
    debug_truthy_values: Vec<String>,
    debug_falsy_values: Vec<String>,
}

#[derive(Deserialize)]
struct SenderCases {
    valid: ValidSender,
    missing_token: ErrorCase,
    missing_target: ErrorCase,
    debug_defaults_false: DebugDefaultCase,
    domains_optional: DomainsOptionalCase,
}

#[derive(Deserialize)]
struct ValidSender {
    env: HashMap<String, String>,
    expected: ExpectedSender,
}

#[derive(Deserialize)]
struct ExpectedSender {
    webex_token: String,
    target: String,
    slug: String,
    domains: Vec<String>,
    debug: bool,
}

#[derive(Deserialize)]
struct ErrorCase {
    env: HashMap<String, String>,
    error_contains: String,
}

#[derive(Deserialize)]
struct DebugDefaultCase {
    env: HashMap<String, String>,
    expected_debug: bool,
}

#[derive(Deserialize)]
struct DomainsOptionalCase {
    env: HashMap<String, String>,
    expected_domains: Vec<String>,
}

#[derive(Deserialize)]
struct BotCases {
    valid: ValidBot,
    missing_domains: ErrorCase,
}

#[derive(Deserialize)]
struct ValidBot {
    env: HashMap<String, String>,
    expected: ExpectedBot,
}

#[derive(Deserialize)]
struct ExpectedBot {
    webex_token: String,
    domains: Vec<String>,
}

#[derive(Deserialize)]
struct ReceiverCases {
    valid: ValidReceiver,
}

#[derive(Deserialize)]
struct ValidReceiver {
    env: HashMap<String, String>,
    expected: ExpectedReceiver,
}

#[derive(Deserialize)]
struct ExpectedReceiver {
    webex_token: String,
    slug: String,
    domains: Vec<String>,
}

fn load_cases() -> ConfigCases {
    let data = fs::read_to_string("../tests/config_cases.json").expect("load config cases");
    serde_json::from_str(&data).expect("parse config cases")
}

fn set_env(env: &HashMap<String, String>) {
    for (key, _) in std::env::vars() {
        if key.starts_with("WGROK_") {
            unsafe { std::env::remove_var(&key); }
        }
    }
    for (k, v) in env {
        unsafe { std::env::set_var(k, v); }
    }
}

fn clear_env() {
    for (key, _) in std::env::vars() {
        if key.starts_with("WGROK_") {
            unsafe { std::env::remove_var(&key); }
        }
    }
}

/// All config tests run sequentially under a single mutex to avoid env var races.
#[test]
fn test_all_config() {
    let _lock = ENV_LOCK.lock().unwrap();
    let cases = load_cases();

    // Sender valid
    set_env(&cases.sender.valid.env);
    let cfg = SenderConfig::from_env().unwrap();
    let exp = &cases.sender.valid.expected;
    assert_eq!(cfg.webex_token, exp.webex_token);
    assert_eq!(cfg.target, exp.target);
    assert_eq!(cfg.slug, exp.slug);
    assert_eq!(cfg.domains, exp.domains);
    assert_eq!(cfg.debug, exp.debug);

    // Sender missing token
    clear_env();
    set_env(&cases.sender.missing_token.env);
    let result = SenderConfig::from_env();
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_lowercase()
        .contains(&cases.sender.missing_token.error_contains.to_lowercase()));

    // Sender missing target
    clear_env();
    set_env(&cases.sender.missing_target.env);
    let result = SenderConfig::from_env();
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_lowercase()
        .contains(&cases.sender.missing_target.error_contains.to_lowercase()));

    // Sender debug defaults false
    clear_env();
    set_env(&cases.sender.debug_defaults_false.env);
    let cfg = SenderConfig::from_env().unwrap();
    assert_eq!(cfg.debug, cases.sender.debug_defaults_false.expected_debug);

    // Sender domains optional
    clear_env();
    set_env(&cases.sender.domains_optional.env);
    let cfg = SenderConfig::from_env().unwrap();
    assert_eq!(cfg.domains, cases.sender.domains_optional.expected_domains);

    // Bot valid
    clear_env();
    set_env(&cases.bot.valid.env);
    let cfg = BotConfig::from_env().unwrap();
    assert_eq!(cfg.webex_token, cases.bot.valid.expected.webex_token);
    assert_eq!(cfg.domains, cases.bot.valid.expected.domains);

    // Bot missing domains
    clear_env();
    set_env(&cases.bot.missing_domains.env);
    let result = BotConfig::from_env();
    assert!(result.is_err());

    // Receiver valid
    clear_env();
    set_env(&cases.receiver.valid.env);
    let cfg = ReceiverConfig::from_env().unwrap();
    assert_eq!(cfg.webex_token, cases.receiver.valid.expected.webex_token);
    assert_eq!(cfg.slug, cases.receiver.valid.expected.slug);
    assert_eq!(cfg.domains, cases.receiver.valid.expected.domains);

    clear_env();
}

#[test]
fn test_debug_parsing() {
    let cases = load_cases();
    for val in &cases.debug_truthy_values {
        assert!(parse_debug(val), "parse_debug({}) should be true", val);
    }
    for val in &cases.debug_falsy_values {
        assert!(!parse_debug(val), "parse_debug({}) should be false", val);
    }
}
